Note: we start with no ``--remove``; BOLT will automatically generate one of the format ``bolt.in_plink_but_not_imputed.FID_IID.XXX.txt`` to be added.

Previous file generation (``longSignal.sh``) restricted to samples who self-identified as 'White British' according to Field 21000 and have very similar genetic ancestry based on a principal components analysis of the genotypes

```
input $CHR
BGEN_PREFIX="/mnt/project/Bulk/Imputation/Imputation from genotype (TOPmed)/ukb21007_c${CHR}_b0_v1"
tar -zxf $DIR_BOLT/BOLT-LMM_v2.5.tar.gz
NUM_PCS=10 # can be increased to 20 if desired

./BOLT-LMM_v2.5/bolt \
    --bfile="/mnt/project/Bulk/Genotype Results/Genotype calls/ukb22418_c21_b0_v2" \
    --phenoFile=phenotypes.${LOCUS}.${REP}.UKB.txt.gz \
    --phenoCol=resid${PHENO}_cropped \
    --covarFile=phenotypes.${LOCUS}.${REP}.UKB.txt.gz \
    --covarCol=cov_SEX_GENETIC --qCovarCol=cov_AGE --qCovarCol=cov_AGE_SQ \
    --qCovarCol=imputed${PHENO} --qCovarCol=imputed${PHENO}_cov_AGE \
    --qCovarCol=imputed${PHENO}_cov_AGE_SQ --qCovarCol=ukbPC{1:$NUM_PCS} \
   --numThreads=$(nproc) \
    --statsFile=/dev/null \
    --bgenFile="$BGEN_PREFIX.bgen" --sampleFile="$BGEN_PREFIX.sample" \
    --bgenMinMAF=0.001 --bgenMinINFO=0.3 \
    --statsFileBgenSnps=$LOCUS.$REP.$PHENO.chr$CHR.covADJ.bgen.stats.gz \
    --verboseStats
 ```
   
Example for MLH3 locus using output from ``longSignal.sh``:

```
LOCUS=chr19:14774675-14774689; REP=AAAG; PHENO=H4
DIR_BOLT=/mnt/project/lohdata/resources/BOLT-LMM/

CHR=14
BGEN_PREFIX="/mnt/project/Bulk/Imputation/Imputation from genotype (TOPmed)/ukb21007_c${CHR}_b0_v1"
bgenix -g "${BGEN_PREFIX}".bgen  -incl-range chr14:74027345-76027345 > MLH3.bgen;

./BOLT-LMM_v2.5/bolt \
    --bfile="/mnt/project/Bulk/Genotype Results/Genotype calls/ukb22418_c21_b0_v2" \
    --remove bolt.in_plink_but_not_imputed.FID_IID.1098.txt \ # file generated by BOLT; then remove statement was added
    --phenoFile=phenotypes.${LOCUS}.${REP}.UKB.txt.gz \
    --phenoCol=resid${PHENO}_cropped \
    --covarFile=phenotypes.${LOCUS}.${REP}.UKB.txt.gz \
    --covarCol=cov_SEX_GENETIC --qCovarCol=cov_AGE --qCovarCol=cov_AGE_SQ \
    --qCovarCol=imputed${PHENO} --qCovarCol=imputed${PHENO}_cov_AGE \
    --qCovarCol=imputed${PHENO}_cov_AGE_SQ --qCovarCol=ukbPC{1:$NUM_PCS} \
   --numThreads=$(nproc) \
    --statsFile=/dev/null \
    --bgenFile=MLH3.bgen --sampleFile="$BGEN_PREFIX.sample" \
    --bgenMinMAF=0.001 --bgenMinINFO=0.3 \
    --statsFileBgenSnps=$LOCUS.$REP.$PHENO.chr$CHR.covADJ.bgen.stats.gz \
    --verboseStats
```

Results in the following top associations:
```
 zcat chr19:14774675-14774689.AAAG.H4.chr14.covADJ.bgen.stats.gz | sort -k10,10g | head  | column -t
SNP  CHR  BP        GENPOS  ALLELE1  ALLELE0  A1FREQ    INFO      CHISQ_LINREG  P_LINREG  BETA      SE
.    14   75014665  0       C        T        0.535186  0.994467  1151.04       2.7E-252  0.180494  0.00532009
.    14   75005406  0       G        A        0.535161  0.994492  1150.46       3.6E-252  0.180446  0.00532
.    14   75016124  0       C        A        0.535208  0.994586  1149.67       5.3E-252  0.180388  0.0053201
.    14   75016125  0       G        T        0.535208  0.994585  1149.68       5.3E-252  0.180388  0.0053201
.    14   75040172  0       G        A        0.53518   0.994636  1149.56       5.6E-252  0.180372  0.00531991
.    14   75015205  0       A        G        0.535189  0.994569  1149.27       6.5E-252  0.180355  0.00532006
.    14   75005100  0       G        A        0.535211  0.994541  1149.15       6.9E-252  0.180345  0.00532004
.    14   75028280  0       T        C        0.535017  0.994539  1148.86       7.9E-252  0.180313  0.00531976
```
